torchtune/modules/transformer.py

```diff
@@ -554,47 +623,86 @@ def __init__(
         self.num_heads = num_heads
         self.head_dim = head_dim
         self.causal_mask = None
-        self.pos = None
         self.num_output_chunks = 0
 
+        # attributes for KV caches during inference
+        self.encoder_max_cache_seq_len = None
+        self.decoder_max_cache_seq_len = None
+
     def set_num_output_chunks(self, num_output_chunks: int) -> None:
         """Used to save memory in combination with :class:`~torchtune.modules.loss.CEWithChunkedOutputLoss`.
         This should be called before the first forward pass, in the recipe."""
         self.num_output_chunks = num_output_chunks
 
-    def setup_caches(self, batch_size: int, dtype: torch.dtype) -> None:
-        """Setup key value caches for attention calculation.
+    def setup_caches(
+        self,
+        batch_size: int,
+        dtype: torch.dtype,
+        *,
+        encoder_max_seq_len: Optional[int] = None,
+        decoder_max_seq_len: Optional[int] = None,
+    ):
+        """
+        Sets up key-value attention caches for inference. For each layer in ``self.layers``:
+        - :class:`~torchtune.modules.TransformerSelfAttentionLayer` will use ``decoder_max_seq_len``.
+        - :class:`~torchtune.modules.TransformerCrossAttentionLayer` will use ``encoder_max_seq_len``.
+        - :class:`~torchtune.modules.fusion.FusionLayer` will use both ``decoder_max_seq_len`` and ``encoder_max_seq_len``.
 
         Args:
             batch_size (int): batch size for the caches.
             dtype (torch.dtype): dtype for the caches.
+            encoder_max_seq_len (Optional[int]): maximum encoder cache sequence length.
+            decoder_max_seq_len (Optional[int]): maximum decoder cache sequence length.
         """
+        has_encoder_layers = any(
+            isinstance(l, TransformerCrossAttentionLayer) or isinstance(l, FusionLayer)
+            for l in self.layers
+        )
```

Comment:
So it might be too much of a sledgehammer to do it this way, but in my mind it's also more explicit about the condition we actually care about.
```suggestion
        has_encoder_layers = any(
            isinstance(m, TransformerCrossAttentionLayer) for m in self.modules()
        )
```